src/averageAges.js

```diff
@@ -55,10 +74,23 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const mothers = people.filter(mom =>
+    people.some(child => child.mother === mom.name));
+
+  const childrenWithMothers = people.filter(child =>
+    (!onlyWithSon && mothers.some(mom => mom.name === child.mother))
+    || (onlyWithSon && child.sex === maleSex
+    && mothers.some(mom => mom.name === child.mother))
+  );
```

Comment:
Now everything is written correctly, but such code is difficult to read and maintain. Break it up into smaller chunks for better readability.